### PR TITLE
Containerapps 4/11 Release [ Supporting MSI for ACA & some bug fixes]

### PR DESCRIPTION
**Pull request by calvinsID**
- P1 & P2 fixes
- MSI feature

### CODE REVIEW

**Comment by yonzhan** [comment]
Containerapps

**Comment by panchagnula** [comment]
@zhoxing-ms can we get your review on this as well. We want to release this version on 4/11 (PST morning is better). Thanks!

File: src/containerapp/azext_containerapp/_help.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_help.py:165]: Is this because we missed this In the existing help?

File: src/containerapp/azext_containerapp/_help.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_help.py:283]: NIT: just to clarify this is user-assigned can we call this 'myUserAssignedId'

Comment (panchagnula) [reply src/containerapp/azext_containerapp/_help.py:283]: This comment is for all examples in help. There is no -n (containerapp name) in the help commands. I would expect this to be required

File: src/containerapp/azext_containerapp/_help.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_help.py:303]: No examples for this help?

File: src/containerapp/azext_containerapp/_help.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_help.py:520]: Curious given we added we update existing values here. What happens with disable? Do we remove existing values?

File: src/containerapp/azext_containerapp/_help.py
Comment (panchagnula) [reply src/containerapp/azext_containerapp/_help.py:165]: Not related  to this change but Line 156-157 of help " az containerapp revision copy -n MyContainerapp -g MyResourceGroup \\
              --cpu 0.75 --memory 1.5Gi" Does the --memory need the value with Units? what happens if one gives just the value? we should clarify this in examples.

**Review by panchagnula — CHANGES_REQUESTED** [review]
Some generic help changes (not updated in this PR)
  text: |
          az containerapp env create -n MyContainerappEnvironment -g MyResourceGroup \\
              --location "Canada Central" 
Is there a command to get the list of regions supported on Container apps? If so  I would not use an location value here as <location of choice> See: command to get list of supported locations. If not change this to East US2 (Canada Central has some capacity constraints I believe)

In general, avoid using "Canada Central" across help examples.

**Comment by panchagnula** [comment]
@haroonf FYI, incase you plan to make the changes here before the Calvin can get to it, Thanks!

**Comment by panchagnula** [comment]
@calvinsID , @haroonf have you tried this example as is on your individual tests?  Mainly this line can cause confusion - since based on a shell this can be parsed differently  --secrets mysecret=secretvalue1 anothersecret="secret value 2" 

I would expect this to be --secrets "sec1=val1 sec2='spaced val2'" Ex. see this https://docs.microsoft.com/en-us/cli/azure/use-cli-effectively#use-quotation-marks-in-arguments. In general I would remove this example from our help, since most users will try to copy paste a help command & expect it to work.

 az containerapp create -n MyContainerapp -g MyResourceGroup \\
              --image my-app:v1.0 --environment MyContainerappEnv \\
              --secrets mysecret=secretvalue1 anothersecret="secret value 2" \\
              --env-vars GREETING="Hello, world" SECRETENV=secretref:anothersecret

**Comment by panchagnula** [comment]
All the examples for --location has the "region name" instead of location. I assume we support both on the API, for consistency purposes use the location name instead of region name (avoiding the need to quotations around the region name) Ex. az group create --name RGlocalContext --location westeurope

File: src/containerapp/azext_containerapp/_params.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_params.py:31]: Update to : "Parameters used as arguments will override same the parameters used in yaml file." Flags can be misleading

File: src/containerapp/azext_containerapp/_params.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_params.py:154]: Should this be 'list' ? i.e containerapp secret list?

File: src/containerapp/azext_containerapp/custom.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/custom.py:214]: Can you add a comment to explain purpose of the helper method. With the definition of the method here , whenever a new property is added to the ContaineApp object we need to update this? Is there a way to make this generic or is this a limitation because we don't use SDK to get the Model object defintion?

File: src/containerapp/azext_containerapp/custom.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/custom.py:244]: Same here add what is this for? especially if we are adding special handling on client side which is different from how the API behaves (making the behavior inconsistent from other clients). Thanks!

File: src/containerapp/azext_containerapp/custom.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/custom.py:249]: Honestly all this special code per argument type is making me nervous. Is this because we are trying to override the yaml arguments with the parameters from the command? Code comments will help here - so that the next person knows when they touch the code. Please add test cases for this in the next PR like we discussed. That will help ensure we have 0 regressions here.

**Review by panchagnula — CHANGES_REQUESTED** [review]
@calvinsID let discuss these YAML specific changes on Monday, while we wait for review from @zhoxing-ms as well & target releasing this 4/11 evening PST. I will make sure to work with @zhoxing-ms

**Comment by panchagnula** [comment]
@calvinsID we also need to update the version on this extension (please do a major version update because the argument vs YAML is a breaking change behavior)

**Comment by panchagnula** [comment]
Lets make the non warning logger.warning - logger.info & add the name of the resource i.e instead of saying updating app say updating app "xyz".

File: src/containerapp/azext_containerapp/_help.py
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_help.py:286]: Please use `--system-assigned` to enable system assigned identity and `--user-assigned` with space separated resource IDs to add user assigned identities.
For more detail, please refer to the guideline [managed_identity_command_guideline.md](https://github.com/Azure/azure-cli/blob/dev/doc/managed_identity_command_guideline.md)

File: src/containerapp/azext_containerapp/_help.py
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_help.py:298]: Same as above

File: src/containerapp/azext_containerapp/_utils.py
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_utils.py:535]: > containerapp_def["properties"]["configuration"]["ingress"]["traffic"]

I want to confirm that you can ensure that there is no case where a parent attribute is `None`?

File: src/containerapp/azext_containerapp/custom.py
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/custom.py:613]: May I ask will the modification of the default value here cause breaking change to users?

**Comment by zhoxing-ms** [comment]
Could you add some tests for those changes?

File: src/containerapp/azext_containerapp/_help.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/_help.py:165]: There are error messages to handle this, I'll add additional helptext to clarify 
image

File: src/containerapp/azext_containerapp/_help.py
Comment (panchagnula) [reply src/containerapp/azext_containerapp/_help.py:286]: @zhoxing-ms thanks for pointing this out - we enabled these commands based on az acr [az acr identity | Microsoft Docs](https://docs.microsoft.com/en-us/cli/azure/acr/identity?msclkid=50244b51b9ba11ec9918d53ebe5b3292&view=azure-cli-latest) Is there a reason these are different from the guidelines?

File: src/containerapp/azext_containerapp/custom.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/custom.py]: Removed this change

**Comment by panchagnula** [comment]
> Could you add some tests for those changes?

@zhoxing-ms we are reverting some major changes & will add tests as part of upcoming changes. Thanks!

File: src/containerapp/azext_containerapp/_help.py
Comment (panchagnula) [reply src/containerapp/azext_containerapp/_help.py:165]: Surface the API error message - we can do this later.

Comment (calvinsID) [reply src/containerapp/azext_containerapp/_help.py:165]: fixed

Comment (calvinsID) [reply src/containerapp/azext_containerapp/_help.py:165]: Made the `--memory` helptext more clear

File: src/containerapp/azext_containerapp/_help.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/_help.py:303]: Added

File: src/containerapp/azext_containerapp/_params.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/_params.py]: Removed yaml changes from this release

File: src/containerapp/azext_containerapp/_help.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/_help.py:283]: Added

File: src/containerapp/azext_containerapp/_params.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/_params.py:154]: Generalized it for `containerapp secret` so that all secret commands with these parameters will have the helptext

File: src/containerapp/azext_containerapp/custom.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/custom.py]: Removed yaml from this release

File: src/containerapp/azext_containerapp/custom.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/custom.py]: Removed yaml for this release

File: src/containerapp/azext_containerapp/_utils.py
Comment (panchagnula) [reply src/containerapp/azext_containerapp/_utils.py]: remove this a part of this release

Comment (calvinsID) [reply src/containerapp/azext_containerapp/_utils.py]: removed

File: src/containerapp/azext_containerapp/_params.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_params.py:78]: @haroonf update this to match the guidelines as well https://github.com/Azure/azure-cli/blob/dev/doc/managed_identity_command_guideline.md#enable-managed-identity-during-resource-creation

File: src/containerapp/azext_containerapp/_help.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_help.py:34]: Did one or both of you try this example btw to check this works?

Comment (calvinsID) [reply src/containerapp/azext_containerapp/_help.py:34]: It works on powershell 7.x and bash 5.x

Comment (panchagnula) [reply src/containerapp/azext_containerapp/_help.py:34]: If this works then I am ok keeping this btw.

File: src/containerapp/azext_containerapp/_help.py
Comment (haroonf) [reply src/containerapp/azext_containerapp/_help.py:277]: If the identity is in the same resource group as the CA, you don't need the entire resource id, otherwise you do

Comment (panchagnula) [reply src/containerapp/azext_containerapp/_help.py:277]: Might be good to another example clarifying for cross sub or cross RG support to include the full resource ID

File: src/containerapp/azext_containerapp/_help.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_help.py:179]: can you check if there are other examples this needs to be updated as well?

**Review by panchagnula — APPROVED** [review]

File: src/containerapp/azext_containerapp/_help.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/_help.py:34]: re-added

File: src/containerapp/azext_containerapp/_help.py
Comment (calvinsID) [reply src/containerapp/azext_containerapp/_help.py]: Fixed by @haroonf

File: src/containerapp/HISTORY.rst
Comment (calvinsID) [inline src/containerapp/HISTORY.rst:6]: Bumped to 0.3.0 since the old containerapp private extension whl files we have been giving customers in the past went up to 0.2.4. We want to avoid confusion by bumping this

File: src/containerapp/azext_containerapp/_help.py
Comment (zhoxing-ms) [reply src/containerapp/azext_containerapp/_help.py]: > Is there a reason these are different from the guidelines?

@panchagnula This is because these commands were created before our guide existed. PR link: https://github.com/Azure/azure-cli/pull/12381
For these old commands, we will also consider migrating them to new commands with the same style in the future

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**
